src/python/pants/backend/native/config/environment.py

```diff
@@ -152,27 +153,35 @@ class CppCompiler(datatype([
     'exe_filename',
     'library_dirs',
     'include_dirs',
+    'extra_args',
 ]), CompilerMixin):
 
-  def get_invocation_environment_dict(self, platform):
-    ret = super(CppCompiler, self).get_invocation_environment_dict(platform).copy()
+  @property
+  def as_invocation_environment_dict(self):
+    ret = super(CppCompiler, self).as_invocation_environment_dict.copy()
 
     ret['CXX'] = self.exe_filename
 
     return ret
 
 
 # TODO(#4020): These classes are performing the work of variants.
-class GCCCCompiler(datatype([('c_compiler', CCompiler)])): pass
+class CToolchain(datatype([('c_compiler', CCompiler), ('c_linker', Linker)])): pass
+
+
+class LLVMCToolchain(datatype([('c_toolchain', CToolchain)])): pass
```

Comment:
Because then I can't figure out how to do e.g. this in `python_native_code.py`:
```
@rule(CToolchain, [Select(PythonNativeCode)])
def select_c_toolchain_for_local_dist_compilation(python_native_code):
  llvm_c_toolchain = yield Get(LLVMCToolchain, NativeToolchain, python_native_code.native_toolchain)
  yield llvm_c_toolchain.c_toolchain


@rule(CppToolchain, [Select(PythonNativeCode)])
def select_cpp_toolchain_for_local_dist_compilation(python_native_code):
  llvm_cpp_toolchain = yield Get(
    LLVMCppToolchain, NativeToolchain, python_native_code.native_toolchain)
  yield llvm_cpp_toolchain.cpp_toolchain
```

because afaik the engine matches on an Exactly constraint when type-checking `@rule` return types. Just now, when trying the edits in your comment and changing `python_native_code.py` to:
```
@rule(CToolchain, [Select(PythonNativeCode)])
def select_c_toolchain_for_local_dist_compilation(python_native_code):
  return Get(LLVMCToolchain, NativeToolchain, python_native_code.native_toolchain)


@rule(CppToolchain, [Select(PythonNativeCode)])
def select_cpp_toolchain_for_local_dist_compilation(python_native_code):
  return Get(LLVMCppToolchain, NativeToolchain, python_native_code.native_toolchain)
```

I get this error message excerpt when running `./pants test tests/python/pants_test/backend/python/tasks:python_native_code_testing -- -vsk test_python_create_platform_specific_distribution`:
```
E         ExecutionError: Received unexpected Throw state(s):
                     E         Computing Select(<pants.backend.python.subsystems.python_native_code.PythonNativeCode object at 0x11141ad90>, =SetupPyNativeTools)
                     E           Noop(No task was available to compute the value.)
```

cc @stuhood: is it correct to say this is happening because of an Exactly type constraint on rule results? Or is there something else I'm missing?

Reply:
I assume by `return Get(...` you mean `yield Get(...`. Gets are only meant to work with yields.

Reply:
Yep. This will give a more useful error post #5788... but currently returning the wrong type from a rule triggers that Noop behaviour.

Reply:
Correct -- I had tried both, and `yield Get(...)` also fails, with a `StopIteration` error, but I had neglected to mention that above.